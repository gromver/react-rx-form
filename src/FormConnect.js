/* eslint-disable no-unused-vars */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'rx-model';
import { merge } from 'rxjs/observable/merge';
import { debounceTime } from 'rxjs/add/operator/debounceTime';
import utils from './utils';

export default class FormConnect extends PureComponent {
  static TICK = 33;

  static propTypes = {
    form: PropTypes.instanceOf(Form).isRequired,
    children: PropTypes.func.isRequired,
    whenForm: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.bool]),
    whenModel: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.bool]),
    debounce: PropTypes.number,
  };

  static defaultProps = {
    debounce: FormConnect.TICK,
    whenForm: false,
    whenModel: false,
  };

  constructor(props, context) {
    super(props, context);

    this.connect(props);
  }

  componentWillReceiveProps(nextProps) {
    if (
      this.props.form !== nextProps.form
      || !utils.compareAttributes(this.props.whenForm, nextProps.whenForm)
      || !utils.compareAttributes(this.props.whenModel, nextProps.whenModel)
      || this.props.debounce !== nextProps.debounce
    ) {
      this.connect(nextProps);
    }
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  subscription;

  connect({ form, whenForm, whenModel, debounce: delay }) {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }

    const stateObservable = form.getStateObservable();
    const attributeObservable = form.getAttributeObservable();
    const validationObservable = form.getValidationObservable();

    const observables = [];

    if (whenModel !== false) {
      const attributes = Array.isArray(whenModel) ? whenModel : [];

      observables.push(attributeObservable.when(attributes));
      validationObservable.when(attributes);

      if (delay > 0) {
        observables.push(validationObservable.debounceTime(delay));
      } else {
        observables.push(validationObservable);
      }
    }

    if (whenForm !== false) {
      const properties = Array.isArray(whenForm) ? whenForm : [];

      observables.push(stateObservable.when(properties));
    }

    this.subscription = merge(...observables).subscribe(() => this.forceUpdate());
  }

  render() {
    return this.props.children(this.props.form);
  }
}
