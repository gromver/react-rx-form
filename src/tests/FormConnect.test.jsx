import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import FormConnect from '../FormConnect';
import TestForm from './forms/TestForm';

describe('Test FormConnect', () => {
  test('FormConnect with no when... props', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    const component = renderer.create(
      <FormConnect form={form} whenForm debounce={0}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('name', 'a');
    form.setStateValue('foo', 'bar');
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(component).toMatchSnapshot();

    mockFn.mockClear();
    await form.validateAttributes('name');
    expect(mockFn).toHaveBeenCalledTimes(0);
    expect(component).toMatchSnapshot();

    mockFn.mockClear();
    form.setStateValue('a', 'a');
    form.setStateValue('b', 'b');
    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(component).toMatchSnapshot();

    mockFn.mockClear();
    form.setAttribute('password', 'abc');
    await form.validateAttributes('password');
    expect(mockFn).toHaveBeenCalledTimes(0);
    expect(component).toMatchSnapshot();
  });

  test('FormConnect with whenForm prop', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    const component = renderer.create(
      <FormConnect form={form} whenForm={['a', 'b']} debounce={0}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('name', 'a');
    form.setStateValue('a', 'a');
    form.setStateValue('b', 'b');
    form.setStateValue('c', 'c');
    form.setStateValue('d', 'd');
    form.setStateValue('e', 'e');
    expect(mockFn).toHaveBeenCalledTimes(2);
    expect(component).toMatchSnapshot();

    mockFn.mockClear();
    form.setAttribute('password', 'abc');
    expect(mockFn).toHaveBeenCalledTimes(0);
    expect(component).toMatchSnapshot();
  });

  test('FormConnect with whenModel prop', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    renderer.create(
      <FormConnect form={form} whenModel={['password']} debounce={0}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('name', 'Paul');
    form.setStateValue('a', 'a');
    form.setStateValue('b', 'b');
    expect(mockFn).toHaveBeenCalledTimes(0);

    mockFn.mockClear();
    form.setAttribute('name', 'John');
    form.setAttribute('password', 'abc');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  test('FormConnect with whenModel and whenForm prop', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    renderer.create(
      <FormConnect form={form} whenModel={['password']} whenForm={['a', 'b']} debounce={0}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('name', 'Paul');
    form.setStateValue('a', 'a');
    form.setStateValue('b', 'b');
    form.setStateValue('c', 'c');
    form.setStateValue('d', 'd');
    expect(mockFn).toHaveBeenCalledTimes(2);

    mockFn.mockClear();
    form.setAttribute('name', 'John');
    form.setAttribute('password', 'abc');
    form.setStateValue('b', 'b');
    form.setStateValue('c', 'c');
    form.setStateValue('d', 'd');
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  test('debounce prop test', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    renderer.create(
      <FormConnect form={form} whenModel={['password']} whenForm={['a', 'b']}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('password', 'Paul');
    form.setStateValue('a', 'a');
    form.setStateValue('b', 'b');
    form.setStateValue('c', 'c');
    form.setStateValue('d', 'd');

    expect(mockFn).toHaveBeenCalledTimes(3);

    mockFn.mockClear();
    form.setAttribute('password', 'abc');
    form.setAttribute('password', 'bvs');
    await form.validate();
    await new Promise(resolve => setTimeout(resolve, FormConnect.TICK));

    expect(mockFn).toHaveBeenCalledTimes(3);
  });

  test('debounce prop test 2', async () => {
    const form = new TestForm();

    const mockFn = jest.fn(f => <span>{f.getFirstError() && f.getFirstError().state}</span>);

    renderer.create(
      <FormConnect form={form} whenModel={['password', 'name']}>{mockFn}</FormConnect>,
    );

    mockFn.mockClear();
    form.setAttribute('password', 'Paul');
    await form.validate();

    await new Promise(resolve => setTimeout(resolve, FormConnect.TICK));
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  test('componentWillReceiveProps test', async () => {
    const wrapper = shallow(<FormConnect
      form={new TestForm()}
      whenModel={['foo', 'bar']}
    >
      {() => false}
    </FormConnect>);

    const mockFn2 = jest.spyOn(wrapper.instance(), 'connect');

    wrapper.setProps({ unknown: 'value' });
    wrapper.setProps({ whenModel: ['foo', 'bar'] });
    expect(mockFn2).toHaveBeenCalledTimes(0);

    wrapper.setProps({ whenModel: ['foo'] });
    wrapper.setProps({ whenModel: ['foo'] });
    wrapper.setProps({ whenForm: ['foo'] });
    wrapper.setProps({ whenForm: ['foo'] });
    wrapper.setProps({ debounce: 100 });
    wrapper.setProps({ form: new TestForm() });
    expect(mockFn2).toHaveBeenCalledTimes(4);
  });
});
