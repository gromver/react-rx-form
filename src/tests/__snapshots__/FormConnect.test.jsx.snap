// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FormConnect FormConnect with no when... props 1`] = `<span />`;

exports[`Test FormConnect FormConnect with no when... props 2`] = `<span />`;

exports[`Test FormConnect FormConnect with no when... props 3`] = `<span />`;

exports[`Test FormConnect FormConnect with no when... props 4`] = `<span />`;

exports[`Test FormConnect FormConnect with whenForm prop 1`] = `<span />`;

exports[`Test FormConnect FormConnect with whenForm prop 2`] = `<span />`;
